ReactWindows/ReactNative.Shared/ReactInstanceManager.cs

```diff
@@ -173,40 +176,70 @@ public Task<ReactContext> CreateReactContextAsync(CancellationToken token)
         /// </returns>
         public async Task<ReactContext> GetReactContextAsync(CancellationToken token)
         {
-            if (!_hasStartedCreatingInitialContext)
+            DispatcherHelpers.AssertOnDispatcher();
+            using (await _lock.LockAsync())
             {
-                throw new InvalidOperationException(
-                    "Use the create method to start initializing the React context.");
-            }
+                if (!_hasStartedCreatingInitialContext)
+                {
+                    throw new InvalidOperationException(
+                        "Use the create method to start initializing the React context.");
+                }
 
-            var contextInitializationTask = _contextInitializationTask;
-            if (contextInitializationTask != null)
-            {
-                return await contextInitializationTask;
+                // By this point context has already been created due to the serialized aspect of context initialization.
+                return _currentReactContext;
             }
-            else
+        }
+
+
+        /// <summary>
+        /// Awaits the currently initializing React context, or creates a new one.
+        /// </summary>
+        /// <param name="token">A token to cancel the request.</param>
+        /// <returns>
+        /// A task to await the React context.
+        /// </returns>
+        public async Task<ReactContext> GetOrCreateReactContextAsync(CancellationToken token)
```

Comment:
>GetOrCreateReactContextAsync [](start = 40, length = 28)

This method used to be purely for convenience, which is why it was in an extension method. Do we really need a "GetOrCreateReactContext"? I think I added it initially for background, so when you resumed to foreground from background the root view didn't trigger a new ReactContext creation. Here's how it's done on Android:
[https://github.com/facebook/react-native/blob/master/ReactAndroid/src/main/java/com/facebook/react/ReactRootView.java#L310-L312](https://github.com/facebook/react-native/blob/master/ReactAndroid/src/main/java/com/facebook/react/ReactRootView.java#L310-L312)

I guess the question is, is this API helpful enough to support background mode that we need to diverge from ReactAndroid? It may just be that this would also be useful in ReactAndroid, but the lack of async/await makes it too complicated.

Reply:
Oh, my main concern was the check of _hasStartedCreatingInitialContext outside the async lock! The method helped me access that variable inside the lock, and made the other 2 more or less test only methods.

---
In reply to: [165950651](https://github.com/Microsoft/react-native-windows/pull/1615#discussion_r165950651) [](ancestors = 165950651)